src/main.cpp

```diff
@@ -936,6 +937,16 @@ bool ContextualCheckTransaction(
         }
     }
 
+    // Rules that apply to NU4 or later:
+    if (NU4Active) {
+        BOOST_FOREACH(const JSDescription& joinsplit, tx.vJoinSplit)
+        {
+            if (joinsplit.vpub_old) {
```

Comment:
I found this comment in attempting to understand this change: https://github.com/zcash/zcash/blob/master/src/primitives/transaction.h#L657

I think that it would be good to (a) document this invariant more prominently where these variables are declared, and (b) perhaps enforce it using a sum type (boost::variant or equivalent).

Reply:
Better yet might be to use Visitor to just make `JSDescription` itself into a sum type, and use only a single private `CAmount` field?